### PR TITLE
Refactor main query method to be easier to parse

### PR DESCRIPTION
**Pull request by dwhenry**
Use better naming and expand implementation to help 
parse the method and understand what it is doing

* Use postgres `<@` method instead of `&&` for 
pre-filtering [https://www.postgresql.org/docs/9.3/static/functions-array.html]
* Better method and variable naming

### CODE REVIEW

File: spec/queries/top_query_spec.rb [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# require 'rails_helper'
```
Comment (mobaig) [inline spec/queries/top_query_spec.rb]: This looks a bit wrong.

File: app/queries/find_exact_match.rb [hunk 2]
```diff
@@ -0,0 +1,39 @@
+class FindExactMatch
+  def initialize(query_field: :tags)
+    raise ArgumentError.new("query_field must be `:tags` or `:links`") unless %i{tags links}.include?(query_field)
+
+    @query_field = query_field.to_sym
+  end
+
+
+  def call(query_hash, document_type)
```
Comment (mobaig) [inline app/queries/find_exact_match.rb]: I'd make query_hash a keyword arg for greater clarity at call sites. But not document type, that's probably fine as it is.

File: app/controllers/subscriber_lists_controller.rb [hunk 3]
```diff
@@ -37,13 +37,13 @@ def build_subscriber_list
   end
 
   def find_subscriber_list
```
Comment (mobaig) [inline app/controllers/subscriber_lists_controller.rb]: With the shorter method names you've introduced I'd be even more inclined to golf this entire method down to 3 lines, for brevity.

File: app/queries/find_exact_match.rb [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    return [] unless query_hash.present?
+
+    subscriber_lists = subscriber_lists_with_all_matching_keys(query_hash)
```
Comment (mobaig) [inline app/queries/find_exact_match.rb]: Appreciate this is an existing method, but `subscriber_lists_where_all_keys_match(query_hash)` reads better.

File: spec/queries/find_exact_match_with_spec.rb [hunk 5]
```diff
@@ -0,0 +1,62 @@
+require 'rails_helper'
```
Comment (mobaig) [inline spec/queries/find_exact_match_with_spec.rb]: spec name

File: app/queries/has_topic_value.rb [hunk 6]
```diff
@@ -0,0 +1,19 @@
+class HasTopicValue
```
Comment (mobaig) [inline app/queries/has_topic_value.rb]: Rename suggestion: `FindByTopic`, with method signature `call(topic:)`.

File: app/queries/with_one_matching_value_for_each_key.rb [hunk 7]
```diff
@@ -0,0 +1,45 @@
+class WithOneMatchingValueForEachKey
```
Comment (mobaig) [inline app/queries/with_one_matching_value_for_each_key.rb]: It's late, and I might be misunderstanding this, but `WithAtLeastOneMatchingValue` is more accurate, I think.

File: app/queries/with_one_matching_value_for_each_key.rb [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  def initialize(query_field: :tags, scope: SubscriberList)
+    raise ArgumentError.new("query_field must be `:tags` or `:links`") unless %i{tags links}.include?(query_field)
+
+    @query_field = query_field.to_sym
+    @scope = scope
+  end
+
+  # Find all lists in which at least one match in the supplied list.
```
Comment (mobaig) [inline app/queries/with_one_matching_value_for_each_key.rb]: `in which at least one value in query_field matches a value in the supplied hash`

File: app/queries/with_one_matching_value_for_each_key.rb [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+        (
+          content_item_tags_or_links[key] & subscriber_list_tags_or_links[key]
+        ).any?
+      end
+    end
+  end
+
+private
+
+  # Return all SubscriberLists which have a subscet of keys from those requested.
```
Comment (mobaig) [inline app/queries/with_one_matching_value_for_each_key.rb]: typo: subscet

File: spec/queries/top_query_spec.rb (hunk 1 above)
Comment (dwhenry) [inline spec/queries/top_query_spec.rb]: deleted

File: app/queries/find_exact_match.rb (hunk 4 above)
Comment (dwhenry) [inline app/queries/find_exact_match.rb]: changed to `subscriber_lists_where_all_keys_present` to keep the right intent

File: app/controllers/subscriber_lists_controller.rb (hunk 3 above)
Comment (dwhenry) [inline app/controllers/subscriber_lists_controller.rb]: done

File: spec/queries/find_exact_match_with_spec.rb (hunk 5 above)
Comment (dwhenry) [inline spec/queries/find_exact_match_with_spec.rb]: done

File: app/queries/has_topic_value.rb (hunk 6 above)
Comment (dwhenry) [inline app/queries/has_topic_value.rb]: done

File: app/queries/with_one_matching_value_for_each_key.rb (hunk 7 above)
Comment (dwhenry) [inline app/queries/with_one_matching_value_for_each_key.rb]: I have changed this to `WithMatchingValuesForAllKeys` its a bit wordy but it's also not a super easy concept

File: app/queries/with_one_matching_value_for_each_key.rb (hunk 9 above)
Comment (dwhenry) [inline app/queries/with_one_matching_value_for_each_key.rb]: fixed

File: app/queries/find_exact_match.rb (hunk 2 above)
Comment (dwhenry) [inline app/queries/find_exact_match.rb]: I understand your reasoning here, however I would prefer to leave this as is for the moment as we will need to add `email_document_supertype` and `government_document_supertype` to the argument list. I would prefer to update the interface when we do that PR. (which is what I am working on today)

File: app/queries/with_one_matching_value_for_each_key.rb (hunk 8 above)
Comment (dwhenry) [inline app/queries/with_one_matching_value_for_each_key.rb]: Made this into an epic.. not sure it is better or not..

Comment (dwhenry) [inline app/queries/with_one_matching_value_for_each_key.rb]:
```
  # Filter by subscriber lists where the keys from the supplied hash (content_item_tags_or_links)
  # are a superset of the keys in corresponding subscriber list field (tags or links) where
  # the array of values for each key (on the subscriber list) have at least one matching value
  # with the array of values from the supplied hash for the matching key.
  # ** This means that not keys from the the supplied hash are required to be matched.
```

File: app/queries/with_matching_values_for_all_keys.rb [hunk 10]
```diff
@@ -1,13 +1,16 @@
-class WithOneMatchingValueForEachKey
+class WithMatchingValuesForAllKeys
```
Comment (mobaig) [inline app/queries/with_matching_values_for_all_keys.rb]: It occurs to me that this class name is probably trying too hard to explain the implementation rather than what this object is for. What if you simply called it something like `MatchedForNotification`, leaving the explanation of the (not exactly simple) query in the comment?

File: app/queries/with_matching_values_for_all_keys.rb [hunk 11]
```diff
@@ -1,13 +1,16 @@
...
     @scope = scope
   end
 
-  # Find all lists in which at least one match in the supplied list.
-  # Note - does not require that all the supplied hash is matched.
+  # Filter by subscriber lists where the keys from the supplied hash (content_item_tags_or_links)
+  # are a superset of the keys in corresponding subscriber list field (tags or links) where
+  # the array of values for each key (on the subscriber list) have at least one matching value
+  # with the array of values from the supplied hash for the matching key.
+  # ** This means that not keys from the the supplied hash are required to be matched.
```
Comment (mobaig) [inline app/queries/with_matching_values_for_all_keys.rb]: Agree that fleshing this out is better. Slightly terser and with a fixed typo:

```
Filter out subscriber lists:
- where the keys from the supplied hash are a superset of the keys in the specified query_field AND
- the key values in the subscriber list have at least one corresponding match in the supplied hash.
Note that this means that not all keys from the the supplied hash are required to be matched.
```

File: app/queries/with_matching_values_for_all_keys.rb (hunk 10 above)
Comment (dwhenry) [inline app/queries/with_matching_values_for_all_keys.rb]: Good shout

File: app/queries/with_matching_values_for_all_keys.rb (hunk 11 above)
Comment (dwhenry) [inline app/queries/with_matching_values_for_all_keys.rb]: done

File: spec/queries/find_exact_match_spec.rb [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+      :subscriber_list,
+      tags: {
+        topics: ["oil-and-gas/licensing"],
+        organisations: ["environment-agency", "hm-revenue-customs"]
+      },
+      document_type: "policy",
+    )
+  end
+
+  it "requires all tag types in the document to be present in the list" do
```
Comment (dwhenry) [inline spec/queries/find_exact_match_spec.rb]: This was copied from the existing tests. It will need to be refactored with the supertype changes so will try and make it clearer when I do that.
